ehrapy/anndata/anndata_ext.py

```diff
@@ -154,63 +154,135 @@ def anndata_to_df(
     if any(column.startswith("ehrapycat") for column in to_obs):
         raise ObsMoveError(
             "Cannot move encoded columns from X to obs. Either undo encoding or remove them from the list!"
         )
-    indices = adata.var_names.isin(to_obs)
-    df = adata[:, indices].to_df()
-    adata._inplace_subset_var(~indices)
-    adata.obs = adata.obs.join(df)
-    updated_num_uns, updated_non_num_uns, num_var = _update_uns(adata, to_obs)
-    # cast numerical values from object
-    adata.obs[num_var] = adata.obs[num_var].apply(pd.to_numeric, errors="ignore", downcast="float")
-    # cast non numerical values from object to either bool (if possible) or category
-    adata.obs = _cast_obs_columns(adata.obs)
-    adata.uns["numerical_columns"] = updated_num_uns
-    adata.uns["non_numerical_columns"] = updated_non_num_uns
+
+    if not set(to_obs).issubset(set(adata.var_names)):
+        raise ValueError(
+            "Cannot move columns with invalid names."
+        )
+
+    if copy:
+        adata = adata.copy()
+
+    if copy_obs:
+        indices = adata.var_names.isin(to_obs)
+        df = adata[:, indices].to_df()
+        adata.obs = adata.obs.join(df)
+
+        num_set = set(adata.uns["numerical_columns"].copy())
+        non_num_set = set(adata.uns["non_numerical_columns"].copy())
+        var_num = []
+        var_non_num = []
+        for var in to_obs:
+            if var in num_set:
+                var_num.append(var)
+            elif var in non_num_set:
+                var_non_num.append(var)
+
+        # cast numerical values from object
+        adata.obs[var_num] = adata.obs[var_num].apply(pd.to_numeric, errors="ignore", downcast="float")
+        # cast non numerical values from object to either bool (if possible) or category
+        adata.obs = _cast_obs_columns(adata.obs)
+
+    else:
+        indices = adata.var_names.isin(to_obs)
+        df = adata[:, indices].to_df()
+        adata._inplace_subset_var(~indices)
+        adata.obs = adata.obs.join(df)
+        updated_num_uns, updated_non_num_uns, num_var = _update_uns(adata, to_obs)
+        # cast numerical values from object
+        adata.obs[num_var] = adata.obs[num_var].apply(pd.to_numeric, errors="ignore", downcast="float")
+        # cast non numerical values from object to either bool (if possible) or category
+        adata.obs = _cast_obs_columns(adata.obs)
+        adata.uns["numerical_columns"] = updated_num_uns
+        adata.uns["non_numerical_columns"] = updated_non_num_uns
 
     if copy:
         return adata
 
 
+def delete_from_obs(adata: AnnData, to_delete: list[str] | str) -> AnnData:
+    """Delete previously copied features from obs.
+    Args:
+        adata: The AnnData object
+        to_delete: The columns to delete from obs
+    Returns:
+        AnnData object with deleted columns from obs.
+    """
+
+    if isinstance(to_delete, str):  # pragma: no cover
+        to_delete = [to_delete]
```

Comment:
why not use a set instead? You are already using subset below and `isin` works with sets as well.